### PR TITLE
[daint-gpu] QE-SIRIUS 6.5.2 for CrayGNU

### PR DESCRIPTION
**Pull request by simonpintarelli**
add files for GNU

### CODE REVIEW

**Comment by lucamar** [comment]
Thanks for the contribution @simonpintarelli: the Intel version of `QuantumESPRESSO-SIRIUS` is already available as a module on Piz Daint (`QuantumESPRESSO-SIRIUS/6.5-rc4-CrayIntel-19.10-cuda-10.1`), in general we provide supported applications with one compiler only.

What's the reason to have a release candidate `rc4`, when the stable release `6.5` of `QuantumESPRESSO` is available https://github.com/QEF/q-e/releases/tag/qe-6.5?

**Comment by simonpintarelli** [comment]
@lucamar Is it ok to have the easybuild files for both GNU/Intel?

It is not QE from the official repo, but from the sirius fork (Anton is fetching the latest commits periodically from main qe):
https://github.com/electronic-structure/q-e-sirius

The tag `rc4` is a hack to account for the updated version of SIRIUS as a dependency of the QuantumESPRESSO-SIRIUS (q-e-sirius) package, since easybuild is not able (to my knowledge) to provide a given version of q-e-sirius at the same time for multiple different versions of a dependency (SIRIUS).

**Comment by lucamar** [comment]
Thanks for your feedback @simonpintarelli: of course we can merge EasyBuild scripts for multiple compilers, but I would keep a single version in production (possibly the most efficient one, which I believe is the Intel build that you have already merged).

About the versioning, I would suggest to avoid using `rc4` and add instead explictly the version of `Sirius` used to build `QuantumESPRESSO` in `versionsuffix`: I will add a suggestion in the file that you have contributed in this pull request, then you can decide about it.

File: easybuild/easyconfigs/q/QuantumESPRESSO/QuantumESPRESSO-SIRIUS-6.5-rc4-CrayGNU-19.10-cuda-10.1.eb [hunk 1]
```diff
@@ -0,0 +1,46 @@
+# created by Anton Kozhevnikov (CSCS)
+# modified by Simon Pintarelli (CSCS)
+easyblock = "ConfigureMake"
+
+name = 'QuantumESPRESSO-SIRIUS'
+version = '6.5-rc4'
+cudaversion = 10.1
+versionsuffix = '-cuda-%s' % cudaversion
```
Comment (lucamar) [inline easybuild/easyconfigs/q/QuantumESPRESSO/QuantumESPRESSO-SIRIUS-6.5-rc4-CrayGNU-19.10-cuda-10.1.eb]:
```suggestion
version = '6.5'
local_siriusversion='6.5.2'
local_cudaversion = 10.1
versionsuffix = '%s-cuda-%s' % (local_siriusversion, local_cudaversion)
```

File: easybuild/easyconfigs/q/QuantumESPRESSO/QuantumESPRESSO-SIRIUS-6.5-rc4-CrayGNU-19.10-cuda-10.1.eb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+homepage = 'https://github.com/electronic-structure/q-e-sirius/'
+description = """ SIRIUS-enabled version of Quantum ESPRESSO """
+
+toolchain = {'name': 'CrayGNU', 'version': '19.10'}
+toolchainopts = {'usempi': True, 'openmp': True, 'pic': True, 'verbose': False, 'opt': True}
+
+sources = ['https://github.com/electronic-structure/q-e-sirius/archive/' + 'v%(version)s-sirius.tar.gz']
+
+dependencies = [
+  ('SIRIUS', '6.5.2', '-cuda-%s' % cudaversion)
```
Comment (lucamar) [inline easybuild/easyconfigs/q/QuantumESPRESSO/QuantumESPRESSO-SIRIUS-6.5-rc4-CrayGNU-19.10-cuda-10.1.eb]:
```suggestion
  ('SIRIUS', '%s' % local_siriusversion, '-cuda-%s' % local_cudaversion)
```

File: easybuild/easyconfigs/s/SIRIUS/SIRIUS-6.5.2-CrayGNU-19.10-cuda-10.1.eb [hunk 3]
```diff
@@ -0,0 +1,41 @@
+# created by Anton Kozhevnikov (CSCS)
+easyblock = 'CMakeMake'
+
+name = 'SIRIUS'
+version = '6.5.2'
+cudaversion = 10.1
+versionsuffix = '-cuda-%s' % cudaversion
```
Comment (lucamar) [inline easybuild/easyconfigs/s/SIRIUS/SIRIUS-6.5.2-CrayGNU-19.10-cuda-10.1.eb]:
```suggestion
local_cudaversion = 10.1
versionsuffix = '-cuda-%s' % local_cudaversion
```

File: easybuild/easyconfigs/s/SIRIUS/SIRIUS-6.5.2-CrayGNU-19.10-cuda-10.1.eb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+builddependencies = [('CMake', '3.14.5', '', True)]
+
+dependencies = [
+  ('GSL', '2.5'),
+  ('libxc', '4.3.4'),
+  ('spglib','1.14.1'),
+  ('ELPA','2019.05.001'),
+  ('cray-hdf5', EXTERNAL_MODULE),
+  ('cudatoolkit/%s.105_3.27-7.0.1.1_4.1__ga311ce7' % cudaversion , EXTERNAL_MODULE),
```
Comment (lucamar) [inline easybuild/easyconfigs/s/SIRIUS/SIRIUS-6.5.2-CrayGNU-19.10-cuda-10.1.eb]:
```suggestion
  ('cudatoolkit/%s.105_3.27-7.0.1.1_4.1__ga311ce7' % local_cudaversion , EXTERNAL_MODULE),
```

File: easybuild/easyconfigs/q/QuantumESPRESSO/QuantumESPRESSO-SIRIUS-6.5-rc4-CrayGNU-19.10-cuda-10.1.eb (hunk 2 above)
Comment (simonpintarelli) [reply easybuild/easyconfigs/q/QuantumESPRESSO/QuantumESPRESSO-SIRIUS-6.5-rc4-CrayGNU-19.10-cuda-10.1.eb]: Thank you, this makes sense.

**Comment by simonpintarelli** [comment]
@lucamar Thank you, I've applied your suggestions. Yes, the Intel version has better efficiency and should be the only production version available.

**Comment by toxa81** [comment]
@lucamar I need to distinguish the versions of modified QE because I regularly merge stuff from the official `develop` branch. They are all QE-6.5, but with some incremental updates. EB needs a tagged archive, so rc1,rc2... are needed in the QE labelling (unless there is a better way to do so).

File: easybuild/easyconfigs/q/QuantumESPRESSO/QuantumESPRESSO-SIRIUS-6.5-rc4-CrayGNU-19.10-cuda-10.1.eb (hunk 1 above)
Comment (toxa81) [reply easybuild/easyconfigs/q/QuantumESPRESSO/QuantumESPRESSO-SIRIUS-6.5-rc4-CrayGNU-19.10-cuda-10.1.eb]: You need to fetch the specific version of QE with SIRIUS from this list of releases https://github.com/electronic-structure/q-e-sirius/releases

**Comment by gppezzi** [comment]
You can also retrieve a specific git commit with eb and use part of the hash for as the version. It works also for cases where you don't have any control over the git repo. 

In this case, since you own the repo, adding a tag seems to be a good solution in my opinion.

**Comment by lucamar** [comment]
The build of `ELPA-2019.05.001-CrayGNU-19.10.eb` failed:
```
test/Fortran/test.F90:110:7:

    use test_util
       1
Fatal Error: Can't open module file 'test_util.mod' for reading at (1): No such file or directory
```
See full log `/scratch/snx3000/jenscscs/TestingEB/daint-gpu/dd2c80c-3285/tmp/eb-6h7on5/easybuild-ELPA-2019.05.001-20200327.193539.vumHf.log` referenced at https://jenkins.cscs.ch/blue/organizations/jenkins/TestingEB/detail/TestingEB/3285/pipeline

**Comment by lucamar** [comment]
Please have a look at #1557 providing `ELPA-2019.11.001-CrayGNU-19.10.eb`, which might be used in this pull request as a dependency instead of `ELPA-2019.05.001-CrayGNU-19.10.eb`.

File: easybuild/easyconfigs/s/SIRIUS/SIRIUS-6.5.2-CrayGNU-19.10-cuda-10.1.eb [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+source_urls = ['https://github.com/electronic-structure/SIRIUS/archive/']
+sources = ['v%(version)s.tar.gz']
+
+builddependencies = [('CMake', '3.14.5', '', True)]
+
+dependencies = [
+  ('GSL', '2.5'),
+  ('libxc', '4.3.4'),
+  ('spglib','1.14.1'),
+  ('ELPA','2019.05.001'),
```
Comment (lucamar) [inline easybuild/easyconfigs/s/SIRIUS/SIRIUS-6.5.2-CrayGNU-19.10-cuda-10.1.eb]:
```suggestion
  ('ELPA','2019.11.001'),
```

**Comment by lucamar** [comment]
Thanks for accepting the proposed change @simonpintarelli: however the build will fail unless the pull request #1557 is merged in the master branch of the repository:
```
ERROR: Missing dependencies: ELPA/2019.11.001-CrayGNU-19.10 (no easyconfig file or existing module found)
```
As soon as it is merged, I will re-trigger the build of this request.

**Comment by simonpintarelli** [comment]
@lucamar Thanks, I'll rebase this PR once #1557 is merged.

**Comment by lucamar** [comment]
retest this please

**Comment by lucamar** [comment]
@simonpintarelli The pull request #1557 has been merged to master: could you please pull the latest changes to your fork? As soon as you commit, the test build should restart automatically

File: easybuild/easyconfigs/s/SIRIUS/SIRIUS-6.5.2-CrayGNU-19.10-cuda-10.1.eb [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  ('cray-hdf5', EXTERNAL_MODULE),
+  ('cudatoolkit/%s.105_3.27-7.0.1.1_4.1__ga311ce7' % local_cudaversion , EXTERNAL_MODULE),
+  ('magma', '2.5.2', '-cuda-%s-mkl' % local_cudaversion),
+  ('SpFFT', '0.9.10', '-cuda-%s-mkl' % local_cudaversion)
+]
+
+separate_build_dir = True
+
+common_prebuildopts = 'module unload cray-libsci && module unload cray-libsci_acc && module load gcc/8.3.0 && module list &&'
```
Comment (lucamar) [inline easybuild/easyconfigs/s/SIRIUS/SIRIUS-6.5.2-CrayGNU-19.10-cuda-10.1.eb:33]:
```suggestion
local_common_prebuildopts = 'module unload cray-libsci && module unload cray-libsci_acc && module load gcc/8.3.0 && module list &&'
```

File: easybuild/easyconfigs/q/QuantumESPRESSO/QuantumESPRESSO-SIRIUS-6.5-rc4-CrayGNU-19.10-cuda-10.1.eb [hunk 7]
```diff
@@ -0,0 +1,47 @@
+# created by Anton Kozhevnikov (CSCS)
+# modified by Simon Pintarelli (CSCS)
+easyblock = "ConfigureMake"
+
+name = 'QuantumESPRESSO-SIRIUS'
+version = '6.5'
```
Comment (lucamar) [inline easybuild/easyconfigs/q/QuantumESPRESSO/QuantumESPRESSO-SIRIUS-6.5-rc4-CrayGNU-19.10-cuda-10.1.eb:6]:
```suggestion
version = '6.5-rc4'
```

File: easybuild/easyconfigs/s/SIRIUS/SIRIUS-6.5.2-CrayGNU-19.10-cuda-10.1.eb [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+prebuildopts = [ common_prebuildopts ]
```
Comment (lucamar) [inline easybuild/easyconfigs/s/SIRIUS/SIRIUS-6.5.2-CrayGNU-19.10-cuda-10.1.eb:35]:
```suggestion
prebuildopts = [ local_common_prebuildopts ]
```

**Comment by lucamar** [comment]
The build failed because the download link https://github.com/electronic-structure/q-e-sirius/archive/v6.5-sirius.tar.gz could not be resolved:
```
== FAILED: Installation ended unsuccessfully (build directory: /tmp/jenscscs/25f4b47gpu/QuantumESPRESSOSIRIUS/6.5/CrayGNU-19.10-SIRIUS-6.5.2-cuda-10.1): build failed (first 300 chars): No file found for source https://github.com/electronic-structure/q-e-sirius/archive/v6.5-sirius.tar.gz
```
See the full log at https://jenkins.cscs.ch/blue/organizations/jenkins/TestingEB/detail/TestingEB/3300/pipeline

The reason is that the filename is `v6.5-rc4-sirius.tar.gz` and the download link uses `%(version)s`, which following my previous suggestion was modified to `6.5` only. Given the reply of @toxa81, I think that you want to keep `-rc4` in the module name as well, therefore I have added a suggestion to go back to the original name, which should fix the download issue.
I have also suggested to change one local variable name adding the prefix `local_` to avoid the warning (`local_common_prebuildopts`).

**Comment by lucamar** [comment]
Thanks for accepting the first suggestion @simonpintarelli: please have a look at the suggestion with the change of the local variable name in the `SIRIUS` recipe too.

**Comment by simonpintarelli** [comment]
@lucamar sorry, I thought I had applied them all.

**Pull request merged**